theano/tensor/basic.py

```diff
@@ -3372,52 +3372,499 @@ def transpose(x, axes=None):
+            ||(PyArray_DESCR(xs)->type_num != PyArray_DESCR(zs)->type_num))
+        { PyErr_SetString(PyExc_NotImplementedError, "type(x), type(y), type(z) are not all the same"); %(fail)s; }
+
+        /* encode the stride structure of _x,_y,_z into a single integer. */
+        unit |= ((Sx[2] == type_size || Nx[2] == 1) ? 0x0 : (Sx[1] == type_size || Nx[1]==1) ? 0x1 : 0x2) << 8;
+        unit |= ((Sy[2] == type_size || Ny[2] == 1) ? 0x0 : (Sy[1] == type_size || Ny[1]==1) ? 0x1 : 0x2) << 4;
+        unit |= ((Sz[2] == type_size || Nz[2] == 1) ? 0x0 : (Sz[1] == type_size || Nz[1]==1) ? 0x1 : 0x2) << 0;
+
+        /* create appropriate strides for malformed matrices that are row or column
+         * vectors, or empty matrices.
+         * In that case, the value of the stride does not really matter, but
+         * some versions of BLAS insist that:
+         *  - they are not smaller than the number of elements in the array,
+         *  - they are not 0.
+         */
+        sx_1 = (Nx[1] > 1) ? Sx[1]/type_size : (Nx[2] + 1);
+        sx_2 = (Nx[2] > 1) ? Sx[2]/type_size : (Nx[1] + 1);
+        sy_1 = (Ny[1] > 1) ? Sy[1]/type_size : (Ny[2] + 1);
+        sy_2 = (Ny[2] > 1) ? Sy[2]/type_size : (Ny[1] + 1);
+        sz_1 = (Nz[1] > 1) ? Sz[1]/type_size : (Nz[2] + 1);
+        sz_2 = (Nz[2] > 1) ? Sz[2]/type_size : (Nz[1] + 1);
+
+        switch (type_num)
+        {
+            case NPY_FLOAT:
+            {
+                float a = 1.0;
+                float b = 0.0;
+                float* x = (float*)PyArray_DATA(xs);
+                float* y = (float*)PyArray_DATA(ys);
+                float* z = (float*)PyArray_DATA(zs);
+                char N = 'N';
+                char T = 'T';
+                int Nz1 = Nz[1], Nz2 = Nz[2], Nx2 = Nx[2];
+                // loop over batch axis
+                for (int i = 0; i < Nz[0]; i++) {
+                    switch(unit)
+                    {
+                        case 0x000: sgemm_(&N, &N, &Nz2, &Nz1, &Nx2, &a, y, &sy_1, x, &sx_1, &b, z, &sz_1); break;
+                        case 0x100: sgemm_(&N, &T, &Nz2, &Nz1, &Nx2, &a, y, &sy_1, x, &sx_2, &b, z, &sz_1); break;
+                        case 0x010: sgemm_(&T, &N, &Nz2, &Nz1, &Nx2, &a, y, &sy_2, x, &sx_1, &b, z, &sz_1); break;
+                        case 0x110: sgemm_(&T, &T, &Nz2, &Nz1, &Nx2, &a, y, &sy_2, x, &sx_2, &b, z, &sz_1); break;
+                        case 0x001: sgemm_(&T, &T, &Nz1, &Nz2, &Nx2, &a, x, &sx_1, y, &sy_1, &b, z, &sz_2); break;
+                        case 0x101: sgemm_(&N, &T, &Nz1, &Nz2, &Nx2, &a, x, &sx_2, y, &sy_1, &b, z, &sz_2); break;
+                        case 0x011: sgemm_(&T, &N, &Nz1, &Nz2, &Nx2, &a, x, &sx_1, y, &sy_2, &b, z, &sz_2); break;
+                        case 0x111: sgemm_(&N, &N, &Nz1, &Nz2, &Nx2, &a, x, &sx_2, y, &sy_2, &b, z, &sz_2); break;
+                        default: PyErr_SetString(PyExc_ValueError, "some matrix has no unit stride"); %(fail)s;
```

Comment:
Since the exact gemm call is dependant on the stride structure, the tests should make sure to test each structure at least once. Is it currently the case?